onnx/optimizer/passes/fuse_bn_into_conv.h

```diff
@@ -0,0 +1,174 @@
+    ONNX_ASSERT(bbn.sizes().size() == 1 && bbn.sizes()[0] == s.sizes()[0]);
+    ONNX_ASSERT(m.sizes().size() == 1 && m.sizes()[0] == s.sizes()[0]);
+    ONNX_ASSERT(var.sizes().size() == 1 && var.sizes()[0] == s.sizes()[0]);
+    ONNX_ASSERT(W.sizes().size() > 2 && W.sizes()[0] == s.sizes()[0]);
+
+    switch(s.elem_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+        eps.sizes().push_back(s.sizes()[0]);
+        eps.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          eps.floats().push_back((float) epsilon);
+        }
+        bc.sizes().push_back(s.sizes()[0]);
+        bc.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          bc.floats().push_back(0.f);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
+          }
+          bc = *bc_iter;
+          ONNX_ASSERT(bc.sizes().size() == 1 && bc.sizes()[0] == s.sizes()[0]);
+        }
+
+        var.add(eps);
+        var.sqrt();
+        s.divide(var);
+        if (!s.is_raw_data()) {
+          const char * ptr = reinterpret_cast<const char *>(s.floats().data());
+          std::string string_rep(ptr, ptr + sizeof(float) * s.floats().size());
+          s.set_raw_data(string_rep);
+        }
+
+        W.scale_by_first_dim(s);
+        bc.subtract(m);
+        bc.multiply(s);
+        bc.add(bbn);
+        break;
+      }
+      case ONNX_NAMESPACE::TensorProto_DataType_DOUBLE:
+      {
+        eps.sizes().push_back(s.sizes()[0]);
+        eps.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_DOUBLE;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          eps.doubles().push_back((double)epsilon);
+        }
+        bc.sizes().push_back(s.sizes()[0]);
+        bc.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_DOUBLE;
+        for (int64_t i = 0; i < eps.sizes()[0]; i++)  {
+          bc.doubles().push_back(0.);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
+          }
+          bc = *bc_iter;
+          ONNX_ASSERT(bc.sizes().size() == 1 && bc.sizes()[0] == s.sizes()[0]);
+        }
+
+        var.add(eps);
+        var.sqrt();
+        s.divide(var);
+        if (!s.is_raw_data()) {
+          const char * ptr = reinterpret_cast<const char *>(s.doubles().data());
+          std::string string_rep(ptr, ptr + sizeof(double) * s.doubles().size());
+          s.set_raw_data(string_rep);
+        }
+        W.scale_by_first_dim(s);
+        bc.subtract(m);
+        bc.multiply(s);
+        bc.add(bbn);
+        break;
+      }
+      default:
+        throw("Incompatible data type");
```

Comment:
@bddppq Do we have Exception types in ONNX?

Reply:
Here shouldn't throw exceptions, because it's just data type that this optimization pass doesn't support (yet), not that the model is invalid.

Reply:
changed handling of different data types so that inputs have the same tensor type and optimization is not run if not float or double